tools/install.sh

```diff
@@ -37,7 +37,7 @@ main() {
         export DEBIAN_FRONTEND=noninteractive
     fi
 
-    if [[ "${SUDO_USER}" = "root" ]]; then
+    if [[ "${USER}" = "root" ]]; then
```

Comment:
This change will remove the possibilty to install with `sudo make install`. Also it will fail the the `${SUDO_USER}` condition on line 45, unless you are using `-u` parameter, e.g. `sudo -u pi make install`. This would break [kiauh](https://github.com/th33xitus/kiauh) for example

Reply:
What do you mean? 

"not_as_root_msg" = DO NOT INSTALL AS ROOT

${SUDO_USER} is root when building in docker via custompios instructions

${USER} is whoever you are invoking sudo as.

this error message in this condition says, DO NOT INSTALL AS ROOT.  so if you are logged in as root, you should not be allowed to install.  is your error message wrong?

Reply:
Logged in as user mryel btw. If you run `sudo -u` it works, yeah, but the wanted installation method is `sudo make install` and not `sudo -u mryel make install`. Logged in as mryel running `make install` will throw the `need_sudo_msg` and running `sudo make install` will throw `not_as_root_msg`. So your implementation doesn't work if you just run the installer on a plain Debian. And I say it again, I'm not talking anything about docker here, it's just about executing `make` to install crowsnest.

Reply:
I agree I shouldn't break the "normal way of installing".  When I tested it on my pi I didn't see a problem so I don't know why you're seeing it.

Instead of closing the PR and ignoring the issue why don't you just let me add a variant of the script / make targets that enables installing via docker?  Seems like you are not interested in finding a way forward by just closing the PR

Reply:
I agree, closing was a bit harsh. But all your current changes are just breaking stuff or adding "useless" `sudo` calls. Also I forgot to mention https://github.com/mainsail-crew/crowsnest#contribute. Please use the conventional commit guidelines, for the single commits and your PR name. We can reopen it if you want to, or you just create a new PR where we start from scratch, your decision 😄 
But I'm wondering, why it worked for you on your pi. Maybe you have accidentally run `sudo -u pi make install` instead of `sudo make install`, as I can guarantee that it won't work with the second one and your changes. The second one should be preferred, as it's a lot easier for linux noobs and some scripts, like kiauh, to use, imho

Reply:
I can tell you that the changes here work 100% for building via docker so the comments on whether sudo as BASE_USER being added as "useless" is inaccurate for this use case. In fact, those additions were found through repeated builds that failed without those changes, one by one.  

I originally had my chances be a completely separate make target and install script but when I tested sudo make install on my pi with these changes I thought it worked so I combined it in the way I did with the original make target.  I'll try again.

Reply:
> so the comments on whether sudo as BASE_USER being added as "useless" is inaccurate for this use case

Ok, that could be but I'm wondering why it needs that 🤔
As I don't have any knowledge yet about the custompios module, how does the module installation get's invoked with the docker image? So what command can be compared to the build process? As I currently can only think of:

1. `make install` as root, this shouldn't need `sudo` as it already got `root/sudo` privileges
2. `make install` as an user, this would indeed need the `sudo` at least for installation and I think the symlink but not for the `sed` commands
3. `sudo make install` as root, this shouldn't be the case and would end up in 1
4. `sudo make install` as an user, this would then not need the `sudo` for the commands, as you got `sudo` privileges
5. `sudo -u user make install` as root or user, this would be the same as 4

So only the second scenario should need it, doesn't it? I'm getting less secure with every command you write.
I sadly currently don't have the time to setup docker and so on, to try it for myself.

Reply:
the user in docker is root, as in, logged in as root.

so when it executes make install, or sudo make install, it fails the "do not run as root" since they are logged in as root.  if you are logged in as root, and run make install, the current HEAD of your mainline should fail.  

this is the core thing I was trying to workaround with this PR.

I think the proper way is to always sudo as BASE_USER since that is supposed to be the default non-root sudoer, which is usually pi.  so I ended up making this PR to run with that.

1. make install when logged in as root -- this fails because if the condition that runs not_as_root_msg
2. make install as a regular user, won't work because of certain things that executes that need sudo as you pointed out
3. sudo make install when logged in as root -- same as 1, fail because of the check that you're not root.
4. sudo make install as a user -- I'm not sure that if you run make as sudo, that all the commands underneath it assume the same context sudo.  isn't that why you have sudo make install, and then inside your actions your also have sudo in some places?
5. sudo -u user make install as root or user changes depending on who is logged in and who you're sudoing as.  the scripts refer to different environment variables that mean logged in user, vs, user executing

try printing ${SUDO_USER} and ${USER} when you are logged in as root, when you sudo as root, and when you are logged in as a non-root sudoer, and try it again with the -u option.  I was surprised when I tried this.  understanding these different situations is what led me to the PR I made

Reply:
> so when it executes make install, or sudo make install, it fails the "do not run as root" since they are logged in as root. if you are logged in as root, and run make install, the current HEAD of your mainline should fail.

That wasn't the point anymore, that I already got 😄 

About the rest, I just meant the sudoing for the rest of the PR. I just related to the core.sh and not the install.sh script there.
> try printing ${SUDO_USER} and ${USER} when you are logged in as root, when you sudo as root, and when you are logged in as a non-root sudoer, and try it again with the -u option

I tried that before writing the review. I can understand your point of using sudo as BASE_USER but we discussed this already internally in another context and it was said, that the `sudo make install` without defining the user, should be preferred for installing, that's why we didn't go with the `-u` option. This is for noobs on linux, as some won't even get what their user is.

So for docker there might should be just another option. I will try to do reproduce it asap, and maybe I can understand the problem then better and think about a solution that won't break anything (only the BASE_USER to USER breaks something for current installation methods).

Without testing I cannot say if adding those `sudo -u` additions make sense. For me and kwad this didn't look like it would make any sense if you run the installation with sudo or as root, as those commands don't have to be executed specifically as the user.

Reply:
i think the best path forward is for me to port these changes back into a separate make target as I had it before i combined it into this PR.  this leaves the normal make install work as you like (unchanged) but provides docker builds a way forward.  i would love for you to try my fork/this PR using the custompios build instructions in docker though so you can see some of the sudo issues exposed in that context which necessitated the changes i made.  

let me know if that sounds good to you and I can hopefully get around to that soon

Reply:
Yes that sounds like the best option. I will try it for myself asap, but I have to do some other stuff atm before I can try that out. Atm, from what you wrote, it sound like a tricky bug to me inside the docker image maybe, as it's working in the buildchain for MainsailOS without any problems. So maybe I can find it or we will just use your newly added make target then. Will you make a new PR then or should I reopen this PR and mark the conversations as resolved for the moment?

Reply:
i'll follow your lead, whatever you prefer

Reply:
Then I would say we let it closed and you open a new PR were you follow the [contribution guidelinese](https://github.com/mainsail-crew/crowsnest#contribute) for your commits. I will test this and write into the issue if I can reproduce it and/or maybe find something obvious wrong with the custompios module or the docker image.

Reply:
> ${SUDO_USER} is root when building in docker via custompios instructions

This is not correct.

If your base image, an rpi os image for example, has a preconfigured user like `pi` and you run something as the root user with `sudo -u pi foobar`, then your `SUDO_USER` is pi not root! root needs no sudo because its the ultimate permission. If `SUDO_USER` is root then there is a misconception.

That said, as I told you in the beginning I need to look at your configs of the whole setup of your custompios sources, because they depend on each other. Furthermore there is no difference in using the Docker Image made by Guysoft or a Github workflow they are both containrized. So even If i want to reproduce I cant because I dont know how you had it configured. Please upload your project to an git repo so that I can take a look.